File: main/src/main/java/com/google/android/apps/muzei/render/MuzeiRendererFragment.java
/*
 * Copyright 2014 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.muzei.render;

import android.annotation.TargetApi;
import android.app.ActivityManager;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.opengl.GLSurfaceView;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.google.android.apps.muzei.util.ImageBlurrer;
import com.google.android.apps.muzei.util.MathUtil;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

public class MuzeiRendererFragment extends Fragment implements
        RenderController.Callbacks,
        MuzeiBlurRenderer.Callbacks {

    private static final String ARG_DEMO_MODE = "demo_mode";
    private static final String ARG_DEMO_FOCUS = "demo_focus";

    private MuzeiView mView;
    private ImageView mSimpleDemoModeImageView;
    private boolean mDemoMode;
    private boolean mDemoFocus;

    public static MuzeiRendererFragment createInstance(boolean demoMode, boolean demoFocus) {
        MuzeiRendererFragment fragment = new MuzeiRendererFragment();
        Bundle args = new Bundle();
        args.putBoolean(ARG_DEMO_MODE, demoMode);
        args.putBoolean(ARG_DEMO_FOCUS, demoFocus);
        fragment.setArguments(args);
        return fragment;
    }

    public MuzeiRendererFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle args = getArguments();
        mDemoMode = args.getBoolean(ARG_DEMO_MODE, false);
        mDemoFocus = args.getBoolean(ARG_DEMO_FOCUS, false);
    }

    @Override
    @TargetApi(Build.VERSION_CODES.KITKAT)
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState) {
        boolean simpleDemoMode = false;
        if (mDemoMode && Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            ActivityManager activityManager = (ActivityManager)
                    getActivity().getSystemService(Context.ACTIVITY_SERVICE);
            if (activityManager.isLowRamDevice()) {
                simpleDemoMode = true;
            }
        }

        if (simpleDemoMode) {
            DisplayMetrics dm = getResources().getDisplayMetrics();
            int targetWidth = dm.widthPixels;
            int targetHeight = dm.heightPixels;
            if (!mDemoFocus) {
                targetHeight = MathUtil.roundMult4(ImageBlurrer.MAX_SUPPORTED_BLUR_PIXELS
                        * 10000 / MuzeiBlurRenderer.DEFAULT_BLUR);
                targetWidth = MathUtil.roundMult4(
                        (int) (dm.widthPixels * 1f / dm.heightPixels * targetHeight));
            }

            mSimpleDemoModeImageView = new ImageView(container.getContext());
            mSimpleDemoModeImageView.setScaleType(ImageView.ScaleType.CENTER_CROP);
            Picasso.with(getActivity())
                    .load("file:///android_asset/starrynight.jpg")
                    .resize(targetWidth, targetHeight)
                    .centerCrop()
                    .into(mSimpleDemoModeLoadedTarget);
            return mSimpleDemoModeImageView;
        } else {
            mView = new MuzeiView(getActivity());
            mView.setPreserveEGLContextOnPause(true);
            return mView;
        }
    }

    private Target mSimpleDemoModeLoadedTarget = new Target() {
        @Override
        public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom loadedFrom) {
            if (!mDemoFocus) {
                // Blur
                ImageBlurrer blurrer = new ImageBlurrer(getActivity());
                Bitmap blurred = blurrer.blurBitmap(bitmap,
                        ImageBlurrer.MAX_SUPPORTED_BLUR_PIXELS, 0);

                // Dim
                Canvas c = new Canvas(blurred);
                c.drawColor(Color.argb(255 - MuzeiBlurRenderer.DEFAULT_MAX_DIM,
                        0, 0, 0));

                bitmap = blurred;
            }

            mSimpleDemoModeImageView.setImageBitmap(bitmap);
        }

        @Override
        public void onBitmapFailed(Drawable drawable) {
        }

        @Override
        public void onPrepareLoad(Drawable drawable) {
        }
    };

    @Override
    public void onHiddenChanged(boolean hidden) {
        super.onHiddenChanged(hidden);
        if (mView == null) {
            return;
        }
        mView.mRenderController.setVisible(!hidden);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mView = null;
    }

    @Override
    public void onPause() {
        super.onPause();
        if (mView == null) {
            return;
        }

        mView.onPause();
    }

    @Override
    public void onResume() {
        super.onResume();
        if (mView == null) {
            return;
        }

        mView.onResume();
    }

    @Override
    public void queueEventOnGlThread(Runnable runnable) {
        if (mView == null) {
            return;
        }

        mView.queueEvent(runnable);
    }

    @Override
    public void requestRender() {
        if (mView == null) {
            return;
        }

        mView.requestRender();
    }

    private class MuzeiView extends GLTextureView {
        private MuzeiBlurRenderer mRenderer;
        private RenderController mRenderController;

        public MuzeiView(Context context) {
            super(context);
            mRenderer = new MuzeiBlurRenderer(getContext(), MuzeiRendererFragment.this);
            setEGLContextClientVersion(2);
            setEGLConfigChooser(8, 8, 8, 8, 0, 0);
            setRenderer(mRenderer);
            setRenderMode(GLSurfaceView.RENDERMODE_WHEN_DIRTY);
            if (mDemoMode) {
                mRenderController = new DemoRenderController(getContext(), mRenderer,
                        MuzeiRendererFragment.this, mDemoFocus);
            } else {
                mRenderController = new RealRenderController(getContext(), mRenderer,
                        MuzeiRendererFragment.this);
            }
            mRenderer.setDemoMode(mDemoMode);
            mRenderController.setVisible(true);
        }

        @Override
        protected void onSizeChanged(int w, int h, int oldw, int oldh) {
            super.onSizeChanged(w, h, oldw, oldh);
            mRenderer.hintViewportSize(w, h);
            mRenderController.reloadCurrentArtwork(true);
        }

        @Override
        protected void onDetachedFromWindow() {
            super.onDetachedFromWindow();
            mRenderController.destroy();
            queueEventOnGlThread(new Runnable() {
                @Override
                public void run() {
                    mRenderer.destroy();
                }
            });
        }
    }
}

File: main/src/main/java/com/google/android/apps/muzei/render/MuzeiRendererFragment.kt
/*
 * Copyright 2014 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.apps.muzei.render

import android.app.ActivityManager
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.opengl.GLSurfaceView
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.lifecycle.observe
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation
import com.bumptech.glide.load.resource.bitmap.CenterCrop
import com.google.android.apps.muzei.settings.EffectsLockScreenOpenLiveData
import com.google.android.apps.muzei.util.ImageBlurrer
import com.google.android.apps.muzei.util.blur
import com.google.android.apps.muzei.util.roundMult4
import java.security.MessageDigest

class MuzeiRendererFragment : Fragment(), RenderController.Callbacks, MuzeiBlurRenderer.Callbacks {

    companion object {

        private const val ARG_DEMO_MODE = "demo_mode"
        private const val ARG_DEMO_FOCUS = "demo_focus"

        private const val TRANSFORMATION_ID = "simpleDemoModeTransformation"
        private val TRANSFORMATION_BYTES = TRANSFORMATION_ID.toByteArray()

        fun createInstance(demoMode: Boolean, demoFocus: Boolean = false): MuzeiRendererFragment {
            return MuzeiRendererFragment().apply {
                arguments = bundleOf(
                        ARG_DEMO_MODE to demoMode,
                        ARG_DEMO_FOCUS to demoFocus)
            }
        }
    }

    private var muzeiView: MuzeiView? = null
    private lateinit var simpleDemoModeImageView: ImageView
    private var demoMode: Boolean = false
    private var demoFocus: Boolean = false

    private val simpleDemoModeTransformation = object : BitmapTransformation() {
        override fun transform(
                pool: BitmapPool,
                bitmap: Bitmap,
                outWidth: Int,
                outHeight: Int
        ) = if (!demoFocus) {
            bitmap.blur(requireContext())?.apply {
                // Dim
                val c = Canvas(this)
                c.drawColor(Color.argb(255 - MuzeiBlurRenderer.DEFAULT_MAX_DIM,
                        0, 0, 0))
            }
        } else {
            bitmap
        }

        override fun equals(other: Any?) = other === this

        override fun hashCode() = TRANSFORMATION_ID.hashCode()

        override fun updateDiskCacheKey(messageDigest: MessageDigest) {
            messageDigest.update(TRANSFORMATION_BYTES)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        demoMode = arguments?.getBoolean(ARG_DEMO_MODE, false) ?: true
        demoFocus = arguments?.getBoolean(ARG_DEMO_FOCUS, false) ?: true
    }

    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View? {
        val activityManager = requireContext().getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager

        if (demoMode && activityManager.isLowRamDevice) {
            val dm = resources.displayMetrics
            var targetWidth = dm.widthPixels
            var targetHeight = dm.heightPixels
            if (!demoFocus) {
                targetHeight = (ImageBlurrer.MAX_SUPPORTED_BLUR_PIXELS * 10000 / MuzeiBlurRenderer.DEFAULT_BLUR)
                        .roundMult4()
                targetWidth = (dm.widthPixels * 1f / dm.heightPixels * targetHeight).toInt().roundMult4()
            }

            simpleDemoModeImageView = ImageView(context).apply {
                scaleType = ImageView.ScaleType.CENTER_CROP
            }
            Glide.with(this)
                    .load("file:///android_asset/starrynight.jpg")
                    .override(targetWidth, targetHeight)
                    .transform(CenterCrop(), simpleDemoModeTransformation)
                    .into(simpleDemoModeImageView)
            return simpleDemoModeImageView
        } else {
            muzeiView = MuzeiView(requireContext())
            return muzeiView
        }
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        muzeiView?.renderController?.visible = !hidden
    }

    override fun onDestroyView() {
        super.onDestroyView()
        muzeiView = null
    }

    override fun onPause() {
        super.onPause()
        muzeiView?.onPause()
    }

    override fun onResume() {
        super.onResume()
        muzeiView?.onResume()
    }

    override fun queueEventOnGlThread(event: () -> Unit) {
        muzeiView?.queueEvent {
            event()
        }
    }

    override fun requestRender() {
        muzeiView?.requestRender()
    }

    private inner class MuzeiView(context: Context) : GLTextureView(context) {
        private val renderer = MuzeiBlurRenderer(getContext(), this@MuzeiRendererFragment, demoMode)
        internal val renderController: RenderController

        init {
            setEGLContextClientVersion(2)
            setEGLConfigChooser(8, 8, 8, 8, 0, 0)
            setRenderer(renderer)
            renderMode = GLSurfaceView.RENDERMODE_WHEN_DIRTY
            preserveEGLContextOnPause = true
            renderController = if (demoMode) {
                DemoRenderController(getContext(), renderer,
                        this@MuzeiRendererFragment, demoFocus)
            } else {
                RealRenderController(getContext(), renderer,
                        this@MuzeiRendererFragment)
            }
            lifecycle.addObserver(renderController)
            if (!demoMode) {
                EffectsLockScreenOpenLiveData.observe(this@MuzeiRendererFragment) {
                    isEffectsLockScreenOpen ->
                    renderController.onLockScreen = isEffectsLockScreenOpen
                }
            }
            renderController.visible = true
        }

        override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
            super.onSizeChanged(w, h, oldw, oldh)
            renderer.hintViewportSize(w, h)
            renderController.reloadCurrentArtwork()
        }

        override fun onDetachedFromWindow() {
            queueEventOnGlThread { renderer.destroy() }
            super.onDetachedFromWindow()
        }
    }
}
